protocol/directory/directory_test.go

```diff
@@ -290,25 +33,74 @@ func TestPoliciesChanges(t *testing.T) {
 	}
 }
 
-func TestSTRHistoryRequestBadRange(t *testing.T) {
-	// create basic test directory
-	d, _ := NewTestDirectory(t, true)
-
-	d.Update()
-
-	res := d.GetSTRHistory(&protocol.STRHistoryRequest{
-		StartEpoch: uint64(4),
-		EndEpoch:   uint64(2)})
+func TestDirectoryKeyLookupInEpochBadEpoch(t *testing.T) {
+	d := NewTestDirectory(t)
+	tests := []struct {
+		name     string
+		userName string
+		ep       uint64
+		want     error
+	}{
+		{"invalid username", "", 0, protocol.ErrMalformedMessage},
+		{"bad end epoch", "Alice", 2, protocol.ErrMalformedMessage},
+	}
+	for _, tt := range tests {
+		res := d.KeyLookupInEpoch(&protocol.KeyLookupInEpochRequest{
+			Username: tt.userName,
+			Epoch:    tt.ep,
+		})
+		if res.Error != tt.want {
+			t.Errorf("Expect ErrMalformedMessage for %s", tt.name)
+		}
+	}
+}
 
-	if res.Error != protocol.ErrMalformedMessage {
-		t.Fatal("Expect ErrMalformedMessage for bad STR history end epoch")
+func TestBadRequestMonitoring(t *testing.T) {
+	d := NewTestDirectory(t)
+
+	tests := []struct {
+		name     string
+		userName string
+		startEp  uint64
+		endEp    uint64
+		want     error
+	}{
+		{"invalid username", "", 0, 0, protocol.ErrMalformedMessage},
+		{"bad end epoch", "Alice", 4, 2, protocol.ErrMalformedMessage},
+		{"out-of-bounds", "Alice", 2, d.LatestSTR().Epoch, protocol.ErrMalformedMessage},
+	}
+	for _, tt := range tests {
+		res := d.Monitor(&protocol.MonitoringRequest{
+			Username:   tt.userName,
+			StartEpoch: tt.startEp,
+			EndEpoch:   tt.endEp,
+		})
+		if res.Error != tt.want {
+			t.Errorf("Expect ErrMalformedMessage for %s", tt.name)
+		}
 	}
+}
 
-	res = d.GetSTRHistory(&protocol.STRHistoryRequest{
-		StartEpoch: uint64(6),
-		EndEpoch:   uint64(d.LatestSTR().Epoch)})
+func TestBadRequestGetSTRHistory(t *testing.T) {
+	d := NewTestDirectory(t)
+	d.Update()
 
-	if res.Error != protocol.ErrMalformedMessage {
-		t.Fatal("Expect ErrMalformedMessage for out-of-bounds STR history")
+	tests := []struct {
+		name    string
+		startEp uint64
+		endEp   uint64
+		want    error
+	}{
+		{"bad end epoch", 4, 2, protocol.ErrMalformedMessage},
+		{"out-of-bounds", 6, d.LatestSTR().Epoch, protocol.ErrMalformedMessage},
+	}
+	for _, tt := range tests {
```

Comment:
I like this structure for our error condition testing. Since so many of our tests follow this structure, is it possible to further refactor this into a generic function that we can use everywhere instead of having to copy/paste this code everywhere (this will make writing future tests easier, too). Here's what I have in mind: We have a generic `TestError` function:

```
func TestError(t *testing.T, tests, testFunction, testRequestType) {
       for _, tt := range tests {
		res := testFunction(newTestRequest(testRequestType, tt))
		if res.Error != tt.want {
			t.Errorf("Expect %s for %s", tt.error, tt.name)
	       }
 	}
 }
```

Which then gets called as `TestError(t, tests, d.GetSTRHistory, protocol.STRType)` in `TestBadRequestGetSTRHistory`. Writing new tests is then only a matter of filling the appropriate test parameter struct and calling `TestError` with the right parameters. What do you think?

Reply:
I'm not sure about this since I think it doesn't really save us much. Besides we can _sometimes_ run sub-tests in parallel (see https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks) using `t.Run`.

Reply:
See my comment https://github.com/coniks-sys/coniks-go/pull/191#discussion_r148081877. If we have cases in which we run parallel sub-tests, I think we could still write two separate functions, one for parallel testing and sequential testing, but still abstract away all the common testing internals.